clang/tools/pattern-rewriter/src/pattern.cpp

```diff
@@ -0,0 +1,497 @@
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "./pattern.h"
+
+#include <optional>
+#include <ostream>
+#include <sstream>
+#include <stdexcept>
+#include <unordered_map>
+#include <variant>
+#include <vector>
+
+namespace pattern {
+
+struct SpacingElement {};
+
+struct LiteralElement {
+  char Value;
+};
+
+struct CodeElement {
+  std::string Name;
+  int SuffixLength = -1;
+};
+
+using Element = std::variant<SpacingElement, LiteralElement, CodeElement>;
+
+using MatchPattern = std::vector<Element>;
+
+struct MatchResult {
+  int Start;
+  int End;
+  std::unordered_map<std::string, std::string> Bindings;
+};
+
+static bool isWhitespace(char Character) {
+  return Character == ' ' || Character == '\t' || Character == '\n';
+}
+
+static bool isRightDelimeter(char Character) {
+  return Character == '}' || Character == ']' || Character == ')';
+}
+
+static int detectIndentation(const std::string &Input, int Start) {
+  int Indentation = 0;
+  int Index = Start - 1;
+  while (Index >= 0 && isWhitespace(Input[Index])) {
+    if (Input[Index] == '\n' || Index == 0) {
+      Indentation = Start - Index - 1;
+      break;
+    }
+    Index--;
+  }
+  return Indentation;
+}
+
+static std::vector<std::string> split(const std::string &Input,
+                                      char Seperator) {
+  std::vector<std::string> Output;
+  const int Size = Input.size();
+  int Index = 0;
+  int Start = 0;
+  while (Index < Size) {
+    char Character = Input[Index];
+    if (Character == Seperator) {
+      Output.push_back(Input.substr(Start, Index - Start));
+      Start = Index + 1;
+    }
+    Index++;
+  }
+  if (Start < Index) {
+    Output.push_back(Input.substr(Start, Index - Start));
+  }
+  return Output;
+}
+
+static std::string join(const std::vector<std::string> Lines,
+                        const std::string &Seperator) {
+  if (Lines.size() == 0) {
+    return "";
+  }
+  std::stringstream OutputStream;
+  const int Count = Lines.size();
+  for (int i = 0; i < Count - 1; i++) {
+    OutputStream << Lines[i];
+    OutputStream << Seperator;
+  }
+  OutputStream << Lines.back();
+  return OutputStream.str();
+}
+
+static std::string trim(const std::string &Input) {
+  const int Size = Input.size();
+  int Index = 0;
+
+  while (Index < Size && isWhitespace(Input[Index])) {
+    Index++;
+  }
+
+  int End = Size - 1;
+  while (End > (Index + 1) && isWhitespace(Input[End])) {
+    End--;
+  }
+
+  return Input.substr(Index, End + 1);
+}
+
+static std::string indent(const std::string Input, int Indentation) {
+  std::vector<std::string> Output;
+  std::stringstream IndentStream;
+  for (int i = 0; i < Indentation; i++) {
+    IndentStream << " ";
+  }
+  const auto Indent = IndentStream.str();
```

Comment:
Is there a clearer way to detect a line with no-whitespace than using trim(Line).size()>0?

Reply:
@LU-JOHN I mean keep using trim() and split() but change the implementation of this 2 functions to return string_view to prevent extra copy in the return time.

@0x12CC  Point taken.
I think we might consider create a manager class to maintain the original string and all the string ref, which can guarantee the life time of all the elements and prevent memory fragment if we want to maintain a vector of all the lines.